content/blog/2017/05/2017-05-12-kubernetes-journey-on-azure.adoc

```diff
@@ -0,0 +1,397 @@
+#
+#   Parameters:
+#     $apiKey:
+#       Contain datadog api key.
+#       Used in secret template
+class profile::kubernetes::resources::datadog (
+    $apiKey = base64('encode', $::datadog_agent::api_key, 'strict')
+  ){
+  include ::stdlib
+  include profile::kubernetes::params
+  require profile::kubernetes::kubectl
+
+  file { "${profile::kubernetes::params::resources}/datadog":
+    ensure => 'directory',
+    owner  => $profile::kubernetes::params::user,
+  }
+
+  profile::kubernetes::apply { 'datadog/secret.yaml':
+    parameters => {
+        'apiKey' => $apiKey
+    },
+  }
+  profile::kubernetes::apply { 'datadog/daemonset.yaml':}
+  profile::kubernetes::apply { 'datadog/deployment.yaml':}
+
+  # As secret's changes do not trigger pods update,
+  # we must reload pods 'manually' in order to use updated secrets.
+  # If we delete a pod defined by a daemonset,
+  # this daemonset will recreate pods automatically.
+  exec { 'Reload datadog pods':
+    path        => ["${profile::kubernetes::params::bin}/"],
+    command     => 'kubectl delete pods -l app=datadog',
+    refreshonly => true,
+    environment => ["KUBECONFIG=${profile::kubernetes::params::home}/.kube/config"] ,
+    logoutput   => true,
+    subscribe   => [
+      Exec['apply datadog/secret.yaml'],
+      Exec['apply datadog/daemonset.yaml'],
+    ],
+  }
+}
+----
+->
+https://github.com/jenkins-infra/jenkins-infra/tree/staging/dist/profile/manifests/kubernetes/resources[More "resources" examples]
+
+
+Let's compare the Puppet approach with the challenges discovered with the
+Jenkins approach.
+
+=== Order matters
+
+With Puppet, it becomes easier to define priorities as
+puppet provides relationship meta parameters and the function 'require' (see
+also:
+https://docs.puppet.com/puppet/4.9/lang_relationships.html[Puppet relationships])
+
+In our Datadog example, we are sure that deployment will respect following order:
+
+  datadog/secret.yaml -> datadog/daemonset.yaml -> datadog/deployment.yaml
+
+Currently, our Puppet code only applies configuration when it detects files’
+changes.  It would be better to compare local files with the cluster
+configurations in order to trigger required updates, but we haven't found a
+good way to improve this yet.
+
+=== Portability
+
+As Puppet is used to configure working environments, it becomes easier to be
+sure that all tools are present and correctly configured.  It's also easier to
+replicate environments and run tests on them with tools like
+http://rspec-puppet.com/[Rpec-puppet], http://serverspec.org/[Serverspec] or
+https://www.vagrantup.com/[Vagrant]
+
+In our Datadog example, we can also easily change the datadog api key depending
+the environment with link:https://docs.puppet.com/hiera/[Hiera].
+
+=== Secret credentials
+
+As we were already using encrypted Hiera with Puppet, we decided to continue to
+use it, making managing secrets for containers very simple.
```

Comment:
It's really not clear what this is referring to. Who is the audience of this post, and are we sure they know what this means?